### PR TITLE
[Framework] Added support for items to parse in mapping

### PR DESCRIPTION
**Pull request by omby8888**
# Description

What - 

A valid mapping can be looked as follows:
```yaml
  - kind: issue
    selector:
      query: .item.name != 'test-item' and .issueType == 'Bug' 
    port:
      itemsToParse: .fields.components
      entity:
        mappings:
          identifier: .item.name
          title: .item.name
          blueprint: '"testBlueprint"'
          properties: {}
          relations:
             issue: .key
```
issue's json from api:
```json
{
    "url": "https://example.com/issue/1",
    "status": "Open",
    "issueType": "Bug",
    "components": [{"name": "Authentication"}, {"name": "Frontend"}],
    "assignee": "user1",
    "reporter": "user2",
    "creator": "user3",
    "priority": "High",
    "created": "2024-03-18T10:00:00Z",
    "updated": "2024-03-18T12:30:00Z",
    "key": "ISSUE-1",
}
```

it will map the component attribute array of each issue got from the api to "testBlueprint" entity, with identifier and parser equals to the component's name

Why - To let integrations map internal array to several entities

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

### CODE REVIEW

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 1]
```diff
@@ -70,22 +70,31 @@ async def _search_as_object(
...
+        async def calculate_raw(data: dict[str, Any]) -> list[dict[str, Any]]:
             should_run = await self._search_as_bool(data, mapping.selector.query)
             if should_run and mapping.port.entity:
-                return await self._search_as_object(
-                    data, mapping.port.entity.mappings.dict(exclude_unset=True)
+                entity_mappings: dict[str, Any] = mapping.port.entity.mappings.dict(
+                    exclude_unset=True
                 )
-            return {}
+                if entity_mappings.get("items_to_parse"):
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:79]:
```suggestion
                if mapping.port.entity.mappings.items_to_parse:
```

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 2]
```diff
@@ -70,22 +70,31 @@ async def _search_as_object(
...
+                    items = await self._search(data, entity_mappings["items_to_parse"])
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:80]:
```suggestion
                    items = await self._search(data, items_to_parse_expression)
```

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 3]
```diff
@@ -70,22 +70,31 @@ async def _search_as_object(
...
+                    if isinstance(items, list):
+                        return [
+                            await self._search_as_object(item, entity_mappings)
+                            for item in items
+                        ]
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:85]: run with async.gather

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py (hunk 2 above)
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:80]: there is data missing from the calculation pls take a look at how the webhook service does it

File: port_ocean/core/handlers/port_app_config/models.py [hunk 4]
```diff
@@ -12,6 +12,7 @@ class EntityMapping(BaseModel):
     team: str | None
     properties: dict[str, str] = Field(default_factory=dict)
     relations: dict[str, str] = Field(default_factory=dict)
+    items_to_parse: str | None = Field(alias="itemsToParse")
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/port_app_config/models.py:15]: pop it one out to the PortResourceConfig

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
some comments

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 5]
```diff
@@ -70,22 +70,31 @@ async def _search_as_object(
     async def _calculate_entities(
         self, mapping: ResourceConfig, raw_data: list[dict[str, Any]]
     ) -> list[Entity]:
-        async def calculate_raw(data: dict[str, Any]) -> dict[str, Any]:
+        async def calculate_raw(data: dict[str, Any]) -> list[dict[str, Any]]:
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:73]: pls note that rn you are also sending the `items_to_parse` to get parsed at line 83
but if ull apply my suggestion to pop this property one level out it should be fine

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (omby8888) [reply port_ocean/core/handlers/entity_processor/jq_entity_processor.py]: done :)

File: port_ocean/core/handlers/port_app_config/models.py (hunk 4 above)
Comment (omby8888) [reply port_ocean/core/handlers/port_app_config/models.py:15]: you sure? in the webhook service from what i saw its in this level and also [here](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/cloud-providers/aws/)

Comment (yairsimantov20) [reply port_ocean/core/handlers/port_app_config/models.py:15]: Yes i try to follow the custom webhook struct

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 6]
```diff
@@ -70,22 +70,43 @@ async def _search_as_object(
...
-            if should_run and mapping.port.entity:
-                return await self._search_as_object(
-                    data, mapping.port.entity.mappings.dict(exclude_unset=True)
-                )
+            if should_run:
+                return await self._search_as_object(data, entity_mappings)
             return {}
 
+        async def calculate_raw(data: dict[str, Any]) -> list[dict[str, Any]]:
+            if mapping.port.entity:
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:82]: no need to validate this anymore as we changed it to be required

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 7]
```diff
@@ -70,22 +70,43 @@ async def _search_as_object(
...
+                entity_mappings: dict[str, Any] = mapping.port.entity.mappings.dict(
+                    exclude_unset=True
+                )
+                if items_to_parse_expression := mapping.port.items_to_parse:
+                    items = await self._search(data, items_to_parse_expression)
+                    if isinstance(items, list):
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:88]: log error anything else as it just ignore extra entities

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 8]
```diff
@@ -70,22 +70,43 @@ async def _search_as_object(
     async def _calculate_entities(
         self, mapping: ResourceConfig, raw_data: list[dict[str, Any]]
     ) -> list[Entity]:
-        async def calculate_raw(data: dict[str, Any]) -> dict[str, Any]:
+        async def calculate_entity(
+            data: dict[str, Any], entity_mappings: dict[str, Any]
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:74]:
```suggestion
            data: dict[str, Any], raw_entity_mappings: dict[str, Any]
```

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 9]
```diff
@@ -70,22 +70,43 @@ async def _search_as_object(
...
-                return await self._search_as_object(
-                    data, mapping.port.entity.mappings.dict(exclude_unset=True)
-                )
+            if should_run:
+                return await self._search_as_object(data, entity_mappings)
             return {}
 
+        async def calculate_raw(data: dict[str, Any]) -> list[dict[str, Any]]:
+            if mapping.port.entity:
+                entity_mappings: dict[str, Any] = mapping.port.entity.mappings.dict(
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:83]:
```suggestion
                raw_entity_mappings: dict[str, Any] = mapping.port.entity.mappings.dict(
```

File: port_ocean/core/handlers/entity_processor/jq_entity_processor.py [hunk 10]
```diff
@@ -70,22 +70,43 @@ async def _search_as_object(
     async def _calculate_entities(
```
Comment (yairsimantov20) [inline port_ocean/core/handlers/entity_processor/jq_entity_processor.py:70]: this function is getting big please move the inner function out with meaningful names

File: CHANGELOG.md [hunk 11]
```diff
@@ -7,6 +7,14 @@ this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 <!-- towncrier release notes start -->
 
+## 0.5.7 (2024-03-18)
+
+
+### Features
+
+- Added ability to map entities from raw array attribute
```
Comment (Tankilevitch) [inline CHANGELOG.md:15]:
```suggestion
- Added the ability to map entities from raw array attribute by introducing `itemsToParse` in the mapping configuration
```

**Review by Tankilevitch — APPROVED** [review]
Very nice job @omby8888 ! 🚀 🌊

**Review by yairsimantov20 — APPROVED** [review]
🌊

File: docs/framework-guides/docs/framework/features/resource-mapping.md [hunk 12]
```diff
@@ -204,18 +248,19 @@ automatically in cases where the target related entity does not exist in the sof
...
-| `resources.[].port.entity.properties`   | dict  | `{}`    | An object of property identifier to JQ expressions that will be used to extract the entity properties.                                                                                                                           |
-| `resources.[].port.entity.relations`    | dict  | `{}`    | An object of relation identifier to JQ expressions that will be used to extract the entity properties.                                                                                                                           |
+| `resources.[].selector.query`\*         | str   |         | A JQ expression that will be used to filter the raw data from the 3rd-party application.                                                                                                                                        |
+| `resources.[].port.itemsToParse`        | str   |         | A JQ expression that will be used to extract multiple entities and apply the mapping on instead of applying it on a single entity.                                                                                              |
```
Comment (MPTG94) [inline docs/framework-guides/docs/framework/features/resource-mapping.md:258]:
```suggestion
| `resources.[].port.itemsToParse`        | str   |         | A JQ expression that will be used to apply the mapping on the items of an array and generate multiple entities from the array items.                                                                                              |
```

**Review by MPTG94 — APPROVED** [review]
Apply my suggestions for docs fixes

**Pull request merged**